namd/src/colvarproxy_namd.C

```diff
@@ -1539,6 +1551,143 @@ int colvarproxy_namd::smp_colvars_loop()
+  //       But what should I do if the user changes the Colvars config dynamically? I wish there is a signal/slot...
+  std::vector<std::vector<cvc_info>> sorted_cvcs(max_depth+1);
+  for (auto it = cvc_depth_map.begin(); it != cvc_depth_map.end(); ++it) {
+    sorted_cvcs[it->second].push_back(cvc_info_map[it->first]);
+  }
+  const bool tf_switch_1 = (cvm::step_relative() > 0) && (!total_forces_same_step());
+  const bool tf_switch_2 = total_forces_same_step();
+  // Calculate the CVCs in parallel. The high-depth ones should be calculated at first.
+  cvm::increase_depth();
+  for (auto it = sorted_cvcs.rbegin(); it != sorted_cvcs.rend(); ++it) {
+    const int numChunks = it->size();
+    const int lowerRange = 0;
+    const int upperRange = numChunks - 1;
+    cvc_info* data = it->data();
+    auto lambda_fn = [data, tf_switch_1, tf_switch_2](int start, int end, void* unused){
+      for (int i = start; i <= end; i++) {
+        std::shared_ptr<colvar::cvc>& cvc_ptr = data[i].cvc_ptr;
+        if (data[i].enabled) {
+          // Follow the order of colvar::calc_cvcs:
+          // 1. calc_cvc_total_force
+          if (data[i].parent_colvar_total_force_calc && tf_switch_1) {
+            cvc_ptr->calc_force_invgrads();
+          }
+          // 2. calc_cvc_values
+          // TODO: Oops! read_data() clears the gradients!
+          //       CVC a has a sub-CVC b, and we run b->calc_gradients at
+          //       first and followed by a->read_data, then the previous gradients are cleared!!
+          //       Without major refactoring, it seems the only way to avoid the problem is
+          //       not to register the atom groups of sub-CVCs to the parent CVC.
+          //       Then I have to rely on modify_children_cvcs_atom_gradients to do backward propagation.
+          //       PLUMED has PLMD::ActionAtomistic and PLMD::ActionWithValue but Colvars has none.
+          cvc_ptr->read_data();
+          cvc_ptr->calc_value();
+          // 3. calc_cvc_gradients
+          if (cvc_ptr->is_enabled(colvardeps::features_cvc::f_cvc_gradient)) {
+            cvc_ptr->calc_gradients();
+            cvc_ptr->calc_fit_gradients();
+            // TODO: Implement debug_gradients for CVC of sub-CVCs
+            if (cvc_ptr->is_enabled(colvardeps::features_cvc::f_cvc_debug_gradient)) {
+              cvc_ptr->debug_gradients();
+            }
+          }
+          // 4. calc_cvc_Jacobians
+          if (data[i].parent_colvar_Jacobian) {
+            cvc_ptr->calc_Jacobian_derivative();
+          }
+          // 5. calc_cvc_total_force
+          if (data[i].parent_colvar_total_force_calc && tf_switch_2) {
+            cvc_ptr->calc_force_invgrads();
+          }
+        }
+      }
+    };
+    CkLoop_Parallelize(numChunks, lowerRange, upperRange,
+                       lambda_fn, NULL, CKLOOP_NONE, NULL);
+  }
+  cvm::decrease_depth();
+  return cvm::get_error();
```

Comment:
### Parallelization and checking the error

When I wrote and test new code, I found that NAMD may not exit cleanly (some other threads were still running) if one thread was exited. I think the implementation of `colvarproxy_namd::error` or at least its use in the parallel region, needs to be revised to use a `std::atomic<bool>` to setup an error "bit" instead of calling `NAMD_die` directly. Then after all threads are finished and joined, we can use `cvm::get_error()` to check the error bit and exit if there is an error.